src/riot-rs-threads/src/arch/cortex_m.rs

```diff
@@ -216,53 +212,43 @@ unsafe extern "C" fn PendSV() {
 /// This function is called in PendSV.
 // TODO: make arch independent, or move to arch
 #[no_mangle]
-unsafe fn sched(old_sp: usize) -> usize {
-    let cs = CriticalSection::new();
-    let next_pid;
-
+unsafe fn sched(old_sp: usize) -> u128 {
     loop {
-        {
-            if let Some(pid) = (&*THREADS.as_ptr(cs)).runqueue.get_next() {
-                next_pid = pid;
-                break;
+        if let Some(res) = critical_section::with(|cs| {
+            let threads = &mut *THREADS.as_ptr(cs);
```

Comment:
Why not use `THREADS.with_mut(|mut threads| { ... } )` here?

Reply:
Because `THREADS.with_mut()` checks for unique access within a CS, which we bypass here (if we are inside PendSV, there can be no other critical section active).

Reply:
Doesn't `THREADS.with_mut()` do the exact thing that we are doing here as well? I.e. create a critical section, borrow the mutex and then execute the closure on it?
What am I missing?

Reply:
Ah, I was missing something, yeah, that sounds good actually.

Reply:
for some reason [that](https://gist.github.com/kaspar030/0a74d4867ae9d48cfe81d56516aa265c)'s 15% slower.

Reply:
That's weird. The only difference that I see is using a raw pointer vs mutable reference to the data in the ref cell...

If that always impacts the performance this strongly, maybe we should directly change the `EnsureOnce::with_mut` implementation to use a raw pointer?

Reply:
yeah. I suggest we get some measurement & testing in place before going with too many changes.

Reply:
... and, postpone this?

Reply:
> yeah. I suggest we get some measurement & testing in place before going with too many changes.

> ... and, postpone this?

Yes, sounds good.